dartagnan/src/main/java/com/dat3m/dartagnan/wmm/utils/graph/immutable/LazyEventGraph.java

```diff
@@ -0,0 +1,259 @@
+package com.dat3m.dartagnan.wmm.utils.graph.immutable;
+
+import com.dat3m.dartagnan.program.event.Event;
+import com.dat3m.dartagnan.wmm.utils.graph.EventGraph;
+import com.dat3m.dartagnan.wmm.utils.graph.mutable.MapEventGraph;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Sets;
+
+import java.util.*;
+import java.util.function.BiConsumer;
+import java.util.function.BiPredicate;
+
+public class LazyEventGraph implements ImmutableEventGraph {
+
+    private final Set<Event> domain;
+    private final Set<Event> range;
+    private final BiPredicate<Event, Event> function;
+    private final boolean empty;
+    private int size = -1;
+
+    public LazyEventGraph(Set<Event> domain, Set<Event> range,
+                          BiPredicate<Event, Event> function) {
+        this.domain = ImmutableSet.copyOf(domain);
+        this.range = ImmutableSet.copyOf(range);
+        this.function = function;
+        this.empty = domain.stream().noneMatch(e1 -> range.stream().anyMatch((e2 -> function.test(e1, e2))));
+    }
+
+    public BiPredicate<Event, Event> getFunction() {
+        return function;
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return empty;
+    }
+
+    @Override
+    public int size() {
+        if (size == -1) {
+            size = domain.stream()
+                    .map(e1 -> range.stream().filter(e2 -> function.test(e1, e2)).count())
+                    .map(s -> Integer.parseInt(s.toString()))
```

Comment:
Why the odd roundtrip through the string representation? Is it just to capture cases where count() gets larger than an int? If so, you can use `Math.toIntExact()`.